custreamz_notebooks/streaming_parsing_aggregations_haproxy.ipynb

```diff
@@ -0,0 +1,294 @@
+    "\n",
+    "    con_cols = ['actconn', 'feconn', 'beconn', 'srvconn', 'retries']\n",
+    "    con_df = clean_df[16].str.split('/')\n",
+    "    for col_id, col_name in enumerate(con_cols):\n",
+    "        clean_df[col_name] = con_df[col_id]\n",
+    "    clean_df.drop_column(16)\n",
+    "    del con_df\n",
+    "\n",
+    "    q_df = clean_df[17].str.split('/')\n",
+    "    clean_df['srv_queue'], clean_df['backend_queue'] = q_df[0], q_df[1]\n",
+    "    clean_df.drop_column(17)\n",
+    "    del q_df\n",
+    "    \n",
+    "    post_parsing_time = int(round(time.time()))\n",
+    "    \n",
+    "    '''\n",
+    "    End of the piecemeal log parsing for HAProxy.\n",
+    "    Simple aggregations to be performed now.\n",
+    "    '''\n",
+    "    \n",
+    "    clean_df['time_backend_response'] = clean_df['time_backend_response'].astype('int')\n",
+    "    avg_backend_response_time = clean_df['time_backend_response'].mean()\n",
+    "    \n",
+    "    post_agg_time = int(round(time.time()))\n",
+    "    \n",
+    "    return \"{0},{1},{2},{3},{4},{5},{6}\".format(num_rows, preprocess_start_time, pre_parsing_time, \\\n",
+    "                                            post_parsing_time, post_agg_time, \\\n",
+    "                                            avg_backend_response_time, size)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's create a Kafka consumer."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Kafka topic to read streaming data from\n",
+    "topic = \"haproxy-logs\"\n",
+    "\n",
+    "#Kafka brokers\n",
+    "bootstrap_servers = 'localhost:9092'\n",
+    "\n",
+    "#Kafka consumer configuration\n",
+    "consumer_conf = {'bootstrap.servers': bootstrap_servers,\n",
+    "                 'group.id': 'custreamz', 'session.timeout.ms': 60000}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now use Streamz to create a Stream from Kafka by polling the topic every 10s."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#If you changed Dask=True, please ensure you have a Dask cluster up and running\n",
+    "stream = Stream.from_kafka_batched(topic, consumer_conf, poll_interval='10s',\n",
+    "                                   npartitions=1, asynchronous=True, dask=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We then use the helper parsing+aggregations function to perform the required operations on each batch polled from Kafka, and write the result into a list."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "final_output = stream.map(haproxy_parsing_aggregations).buffer(100000).gather().sink_to_list()"
```

Comment:
Since you are using a buffer size of 100,000 here it would be best to produce more than 5,000 random HAProxy messages to Kafka in the "random generator" notebook. Otherwise the prints of "final_output" are likely to always be the same since the input is so small.

Reply:
Fixed